File: solutions/elixir/048-unique-paths.ex
# Problem 48: Unique Paths (LeetCode 62)
# Difficulty: Med
# Language: Elixir
# 
defmodule UniquePaths do
  def unique_paths(m, n) do
    k = min(m - 1, n - 1)
    Enum.reduce(0..(k - 1), 1, fn i, acc ->
      div(acc * (m + n - 2 - i), i + 1)
    end)
  end
end

File: solutions/haskell/048-unique-paths.hs
{-
  Problem 48: Unique Paths (LeetCode 62)
  Difficulty: Med
  Language: Haskell
-}
uniquePaths :: Int -> Int -> Int
uniquePaths m n =
  let k = min (m-1) (n-1)
  in foldl (\acc i -> acc * (m + n - 2 - i) `div` (i + 1)) 1 [0..k-1]
